### PR TITLE
Update quaternion ekf

### PR DESCRIPTION
**Pull request by eisoku9618**
- クォータニオンを正規化する部分でバグがあったので直しました
- ジャイロセンサの信頼度を表現できるようにしました
- （リファクタリングしました）

デフォルトでは何の影響もありませんので，よろしくお願いします．

### CODE REVIEW

**Comment by k-okada** [comment]
テストコードできるとうれしいんだけどな。
％ 企業はテストコードからつくるようですよ

--
◉ Kei Okada

**Comment by k-okada** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.jsk.imi.i.u-tokyo.ac.jp:8080/job/hrpsys-qnx/2565/
Test PASSed.

File: rtc/KalmanFilter/EKFilter.h [hunk 1]
```diff
@@ -7,134 +8,90 @@
 
 class EKFilter {
 public:
-  EKFilter() {
+  EKFilter()
+    : P(Eigen::Matrix<double, 7, 7>::Identity() * 0.1),
```
Comment (haudren) [inline rtc/KalmanFilter/EKFilter.h]: Please define a matrix7d for clarity

File: rtc/KalmanFilter/EKFilter.h [hunk 2]
```diff
@@ -7,134 +8,90 @@
...
+                          const Eigen::Vector3d& drift) const {
     /* x_a_priori = f(x, u) */
-    Eigen::Matrix<double, 7, 1> ret;
     Eigen::Matrix<double, 4, 1> q_a_priori;
     Eigen::Vector3d gyro_compensated = gyro - drift;
-    q_a_priori = q + dt / 2 * calcOmega(gyro_compensated) * q;
-    ret.block<4, 1>(0, 0) = q_a_priori;
-    ret.block<3, 1>(4, 0) = drift;
-    return ret;
+    Eigen::Matrix<double, 4, 4> omega;
```
Comment (haudren) [inline rtc/KalmanFilter/EKFilter.h]: Please use a eigen::matrix4d

File: rtc/KalmanFilter/EKFilter.h [hunk 3]
```diff
@@ -142,61 +99,50 @@ class EKFilter {
...
-      Eigen::Matrix<double, 7, 1> x = getx();
-      Eigen::Quaternion<double> q = Eigen::Quaternion<double>(x[0], x[1], x[2], x[3]);
-      hrp::Vector3 eulerZYX = q.toRotationMatrix().eulerAngles(2,1,0);
-      rpy(2) = eulerZYX(0);
-      rpy(1) = eulerZYX(1);
-      rpy(0) = eulerZYX(2);
+    prediction(gyro);
+    correction(acc);
+    /* ekf_filter.printAll(); */
+    Eigen::Quaternion<double> q = Eigen::Quaternion<double>(x[0], x[1], x[2], x[3]);
```
Comment (haudren) [inline rtc/KalmanFilter/EKFilter.h]: Can you rather use the quarternion constructor that takes a vector 4 as argument?

File: rtc/KalmanFilter/EKFilter.h [hunk 4]
```diff
@@ -7,134 +8,90 @@
...
       w[2],  w[1], -w[0],     0;
-    return omega;
   }
 
-  Eigen::Matrix<double, 7, 1> calcPredictedState(Eigen::Matrix<double, 4, 1> q,
-                                                 Eigen::Vector3d gyro,
-                                                 Eigen::Vector3d drift,
-                                                 const double& dt) {
+  void calcPredictedState(Eigen::Matrix<double, 7, 1>& _x_a_priori,
+                          const Eigen::Matrix<double, 4, 1>& q,
```
Comment (haudren) [inline rtc/KalmanFilter/EKFilter.h]: Please use eigen::vector4d

File: rtc/KalmanFilter/EKFilter.h [hunk 5]
```diff
@@ -142,61 +99,50 @@ class EKFilter {
...
-    Eigen::Matrix<double, 7, 3> K;
-    Eigen::Vector3d y;
-    /* need to normalize q_a_priori ? */
-    y = calcMeasurementResidual(z, vel_ref, acc_ref, angular_rate_ref, q_a_priori);
-    H = calcH(q_a_priori);
-    S = H * P_a_priori * H.transpose() + R;
-    K = P_a_priori * H.transpose() * S.inverse();
+    Eigen::Vector3d y = calcMeasurementResidual(z, q_a_priori);
+    calcH(H, q_a_priori);
+    Eigen::Matrix<double, 3, 3> S = H * P_a_priori * H.transpose() + R;
```
Comment (haudren) [inline rtc/KalmanFilter/EKFilter.h]: Please use a matrix3d

File: rtc/KalmanFilter/EKFilter.h [hunk 6]
```diff
@@ -7,134 +8,90 @@
...
-      - dt / 2 * q[0],   dt / 2 * q[3], - dt / 2 * q[2],
-      - dt / 2 * q[3], - dt / 2 * q[0],   dt / 2 * q[1],
-      dt / 2 * q[2], - dt / 2 * q[1], - dt / 2 * q[0];
-    F.block<3, 4>(0, 4) = Eigen::Matrix<double, 3, 4>::Zero();
-    F.block<3, 3>(4, 4) = Eigen::Matrix<double, 3, 3>::Identity();
-    return F;
+      + dt / 2 * q[1], + dt / 2 * q[2], + dt / 2 * q[3],
+      - dt / 2 * q[0], + dt / 2 * q[3], - dt / 2 * q[2],
+      - dt / 2 * q[3], - dt / 2 * q[0], + dt / 2 * q[1],
+      + dt / 2 * q[2], - dt / 2 * q[1], - dt / 2 * q[0];
```
Comment (haudren) [inline rtc/KalmanFilter/EKFilter.h]: Could you rather set F entirely then multiply it by dt/2 for conciseness?

**Comment by haudren** [comment]
Hello,
I made a few remarks above, I hope they will not come out as too blunt. Could you fix these style issues? They really hurt readability.

Otherwise there is maybe room for improvement and refactoring in the computation of the covariance matrix. 

Thanks

**Comment by haudren** [comment]
Oh and also, It would be best to squash some commits before merging, so that the commits that are obviously just correcting some mistakes do not appear in the final tree (e.g. 182f8c9, 1d24bd7...) and only the ones adding functionality are preserved.

ご迷惑にかけてしますが、よろしくお願いします。

**Comment by eisoku9618** [comment]
@haudren Thank you for reviewing !
I've fixed all the issues you mentioned.

**Comment by k-okada** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.jsk.imi.i.u-tokyo.ac.jp:8080/job/hrpsys-qnx/2566/
Test PASSed.

**Pull request merged**